bundles/org.openhab.binding.bmwconnecteddrive/README.md

```diff
@@ -0,0 +1,849 @@
+    Text      item=i3Timer1Departure          
+    Setpoint  item=i3Timer1DepartureHour maxValue=23 step=1 icon="time"
+    Setpoint  item=i3Timer1DepartureMinute maxValue=55 step=5 icon="time"
+    Text      item=i3Timer1Days               
+    Switch    item=i3Timer1DayMon            
+    Switch    item=i3Timer1DayTue            
+    Switch    item=i3Timer1DayWed            
+    Switch    item=i3Timer1DayThu            
+    Switch    item=i3Timer1DayFri            
+    Switch    item=i3Timer1DaySat            
+    Switch    item=i3Timer1DaySun            
+    Switch    item=i3Timer1Enabled            
+    Text      item=i3Timer2Departure          
+    Setpoint  item=i3Timer2DepartureHour maxValue=23 step=1 icon="time"
+    Setpoint  item=i3Timer2DepartureMinute maxValue=55 step=5 icon="time"
+    Text      item=i3Timer2Days               
+    Switch    item=i3Timer2DayMon            
+    Switch    item=i3Timer2DayTue            
+    Switch    item=i3Timer2DayWed            
+    Switch    item=i3Timer2DayThu            
+    Switch    item=i3Timer2DayFri            
+    Switch    item=i3Timer2DaySat            
+    Switch    item=i3Timer2DaySun            
+    Switch    item=i3Timer2Enabled            
+    Text      item=i3Timer3Departure          
+    Setpoint  item=i3Timer3DepartureHour maxValue=23 step=1 icon="time"
+    Setpoint  item=i3Timer3DepartureMinute maxValue=55 step=5 icon="time"
+    Text      item=i3Timer3Days               
+    Switch    item=i3Timer3DayMon            
+    Switch    item=i3Timer3DayTue            
+    Switch    item=i3Timer3DayWed            
+    Switch    item=i3Timer3DayThu            
+    Switch    item=i3Timer3DayFri            
+    Switch    item=i3Timer3DaySat            
+    Switch    item=i3Timer3DaySun            
+    Switch    item=i3Timer3Enabled            
+    Switch    item=i3OverrideEnabled            
+    Text      item=i3OverrideDeparture          
+    Setpoint  item=i3OverrideDepartureHour maxValue=23 step=1 icon="time"
+    Setpoint  item=i3OverrideDepartureMinute maxValue=55 step=5 icon="time"
+  } 
+  Frame label="Last Destinations" {    
+    Text  item=i3DestName                 
+    Text  item=i3DestLocation                                                                                   
+  }  
+  Frame label="Image Properties" {
+    Text    item=i3ImageViewport
+    Text    item=i3ImageSize 
+  } 
+}
+```
+
+### Action example
+
+```
+  val profile = getActions("chargeprofile", "bmwconnecteddrive:bev_rex:user:i3")
+  val now = ZonedDateTime.now.toLocalTime
+  profile.setChargingMode("DELAYED_CHARGING")
+  profile.setTimer1Departure(now.minusHours(2))
+  profile.setTimer1Days(java.util.Set())
+  profile.setTimer1Enabled(true)
+  profile.setTimer2Enabled(false)
+  profile.setTimer3Enabled(false)
+  profile.setPreferredWindowStart(now.minusHours(6))
+  profile.setPreferredWindowEnd(now.minusHours(2))
+  profile.send()
```

Comment:
Why isn't the command sent after each action automatically? That can lead to race conditions if multiple rules are executed simultaneously.

Reply:
Sending a command including delivery to the car takes significant time - 30 seconds is not unusual. In addition to this there is no way to send individual properties of the charging profile, they can only be sent all at once. To ensure that parallel access will not result in a mixture of edits each actions instance does work on it's own copy of the current charging-profile. The global state of the thing is updated when the command is sent and delivery is confirmed by the api.
Of course there still is a race-condition as other rules or users do not see a parallel edit unless it is confirmed as delivered. To avoid such races that the binding might implement locking internal state but I don't think that is the right way to go as there is no common agreement how that should behave. How shall that 'channel is locked'-state made visible to the user and rules? Introduce more channels for that? I think if avoidance of race-conditions caused by parallel edits were a formal requirement there should be some locking-concept in the OH-basis in the first place.

Reply:
I think your current implementation is reasonable under these circumstances.